pkg/controller/registry/reconciler.go

```diff
@@ -0,0 +1,360 @@
+package registry
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/operator-framework/operator-lifecycle-manager/pkg/api/apis/operators/v1alpha1"
+	"github.com/operator-framework/operator-lifecycle-manager/pkg/lib/operatorclient"
+	"github.com/operator-framework/operator-lifecycle-manager/pkg/lib/operatorlister"
+	"github.com/operator-framework/operator-lifecycle-manager/pkg/lib/ownerutil"
+	"github.com/sirupsen/logrus"
+	"k8s.io/api/core/v1"
+	rbacv1 "k8s.io/api/rbac/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/util/intstr"
+)
+
+var timeNow = func() metav1.Time { return metav1.NewTime(time.Now().UTC()) }
+
+// catalogSourceDecorator wraps CatalogSource to add additional methods
+type catalogSourceDecorator struct {
```

Comment:
Why not just use pure functions instead of embedding a struct?

Reply:
It's mostly a style issue; I personally think writing compositions of objects in go ends up more readable than writing pure functions. 

That said, all of the methods I defined on the decorator require state from the CatalogSource. Wrapping it in a (private) struct is equivalent to defining a set of functions that all take the catalog source as a first argument.

The way I'm using the struct here is way more like you'd use a State Monad to thread common state throughout pure function calls.